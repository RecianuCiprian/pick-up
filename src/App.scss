@import "./styles/_helpers/mixins";
@import "./styles/global/variables";

.rc-app {
  display: flex;
  align-items: center;
  justify-content: center;
  height: 100vh;

  &__pick_up {
    @include mds-breakpoint('phone') {
      width: 100%;
    }

    @include mds-breakpoint('to-tablet') {
      width: 100%;
    }

    @include mds-breakpoint('from-laptop') {
      width: 50%;
    }

    @include mds-breakpoint('from-tablet') {
      width: 50%;
    }

    height: 50%;
    background-color: $color-search-location;
    display: flex;
    flex-direction: column;
    align-items:center;

    &__title{
      margin-top: 50px;
      @include mds-font-size(30px);
      font-weight: bold;
    }

    &--position {
      display: flex;
      flex-direction: column;

      label {
        padding-bottom: 15px;
      }

      input {
        height: 10px;
        margin: 0;
        border-radius: 0;
        background: $color-input;
        border: 1px solid $color-input_border;
        font-size: 0.813em;
        padding: 10px;
      }
    }
  }

  &__dropdown {
    margin-top: 60px;
    position: relative;
    display: inline-block;
    @include mds-breakpoint('phone') {
      width: 80%;
    }

    @include mds-breakpoint('to-tablet') {
      width: 50%;
    }
    @include mds-breakpoint('from-tablet') {
      width: 60%;
    }


    @include mds-breakpoint('from-laptop') {
      width: 50%;
    }
  }

  &__dropdown-content {
    display: none;
    position: absolute;
    background-color: $color-white;
    width: 100%;
    overflow: auto;
    box-shadow: 0px 8px 16px 0px rgba(0, 0, 0, 0.2);
    z-index: 1;
  }

  .show {
    display: block;
  }

  &__results{
    border-bottom: 0.5px solid $color-result;
    padding: 6px 0 6px 20px;
    cursor: pointer;

    &__title{
      @include mds-font-size(20px);
      font-weight: bold;
    }

    &__info{
      @include mds-font-size(16px);
    }
    &:hover{
      background-color: $color-result_hover;
    }
  }
  &__no-results{
    padding: 6px 0 6px 20px;
  }

  &__error{
    padding-top: 20px;
    color: $color-error;
    font-weight: bold;
  }
}
