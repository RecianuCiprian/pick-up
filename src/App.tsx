import * as React from 'react';
import './App.scss';
// @ts-ignore
import { debounce } from 'throttle-debounce';
import cx from 'classnames';
import { IAppState } from './AppI';

class App extends React.Component<any, IAppState> {
    private readonly emitChangeDebounced: any;

    constructor(props: any) {
        super(props);
        this.state = {
            results: {
                docs: [],
                numFound: 0,
            },
            err: '',
        };
        this.emitChangeDebounced = debounce(500, this.handleChangeDebounce);
    }

    public render() {
        const {results: {docs, numFound}, err} = this.state;
        const classDropdown = cx('rc-app__dropdown-content', {'show': docs});

        return (
            <div className="rc-app">
                <div className="rc-app__pick_up">
                    <div className="rc-app__pick_up__title">Where are you going?</div>
                    <div className="rc-app__dropdown">
                        <div className="rc-app__pick_up--position">
                            <label htmlFor="location">Pick-up Location</label>
                            <input
                                type="text"
                                placeholder="city, airport, station, region and district..."
                                onChange={this.onChange}
                                id="location"
                                aria-label="Pick Up Location"
                            />
                        </div>
                        <div className={classDropdown}>
                            {docs && docs.map((doc: any) => {
                                if (!numFound) {
                                    return (
                                        <div className="rc-app__no-results">
                                            {doc.name}
                                        </div>
                                    )
                                }
                                const info = doc.city ? `${doc.city}, ${doc.country}` : `${doc.country}`;
                                return (
                                    <div className="rc-app__results" key={doc.bookingId}>
                                        <div className="rc-app__results__title">{doc.name}</div>
                                        <div className="rc-app__results__info">{info}</div>
                                    </div>
                                )

                            })}
                        </div>
                    </div>
                    {err && (
                        <div className="rc-app__error">
                            {err}
                        </div>
                    )}
                </div>
            </div>
        );
    }

    private onChange = (e: any) => {
        const {value} = e.target;
        this.emitChangeDebounced(value)
    };

    private handleChangeDebounce(value: string) {
        if (value.length > 1) {
            this.fetchData(value, 6);
        } else {
            this.setState({
                results: {
                    docs: [],
                    numFound: 0,
                },
            })
        }
    }

    private fetchData(input = '', requiredNumber = 0) {
        fetch(`https://cors.io/?https://www.rentalcars.com/FTSAutocomplete.do?solrIndex=fts_en&solrRows=${requiredNumber}&solrTerm=${input}`)
            .then((data) => {
                data.json().then((responseData) => {
                    const {results = {}} = responseData;
                    this.setState({
                        results,
                    })
                });
            })
            .catch((err) => {
                console.error(err);
                this.setState({
                    err: 'An error has occurred',
                })
            })
    }
}

export default App;
