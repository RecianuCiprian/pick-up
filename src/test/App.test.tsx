import * as React from 'react';
import * as renderer from 'react-test-renderer';
import * as enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';

enzyme.configure({adapter: new Adapter()});
import App from '../App';

test('render app with no data', () => {
    const component = renderer.create(
        <App/>,
    );
    expect(component.toJSON()).toMatchSnapshot();
});
