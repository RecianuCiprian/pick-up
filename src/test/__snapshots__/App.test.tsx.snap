// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render app with no data 1`] = `
<div
  className="rc-app"
>
  <div
    className="rc-app__pick_up"
  >
    <div
      className="rc-app__pick_up__title"
    >
      Where are you going?
    </div>
    <div
      className="rc-app__dropdown"
    >
      <div
        className="rc-app__pick_up--position"
      >
        <label
          htmlFor="location"
        >
          Pick-up Location
        </label>
        <input
          aria-label="Pick Up Location"
          id="location"
          onChange={[Function]}
          placeholder="city, airport, station, region and district..."
          type="text"
        />
      </div>
      <div
        className="rc-app__dropdown-content show"
      />
    </div>
    
  </div>
</div>
`;
